File: go/mjpeg/rtp_packetizer_test.go
package mjpeg

import (
	"bytes"
	"encoding/binary"
	"testing"
)

// TestNewRTPPacketizer tests packetizer initialization
func TestNewRTPPacketizer(t *testing.T) {
	tests := []struct {
		name     string
		ssrc     uint32
		mtu      int
		wantMTU  int
		wantMaxPayload int
	}{
		{
			name:           "default MTU",
			ssrc:           0x12345678,
			mtu:            0,
			wantMTU:        DefaultMTU,
			wantMaxPayload: DefaultMTU - RTPHeaderSize - JPEGHeaderSize,
		},
		{
			name:           "custom MTU",
			ssrc:           0x87654321,
			mtu:            1500,
			wantMTU:        1500,
			wantMaxPayload: 1500 - RTPHeaderSize - JPEGHeaderSize,
		},
		{
			name:           "small MTU",
			ssrc:           0xAABBCCDD,
			mtu:            500,
			wantMTU:        500,
			wantMaxPayload: 500 - RTPHeaderSize - JPEGHeaderSize,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			p := NewRTPPacketizer(tt.ssrc, tt.mtu)

			if p.ssrc != tt.ssrc {
				t.Errorf("SSRC = %x, want %x", p.ssrc, tt.ssrc)
			}

			if p.mtu != tt.wantMTU {
				t.Errorf("MTU = %d, want %d", p.mtu, tt.wantMTU)
			}

			if p.maxPayloadSize != tt.wantMaxPayload {
				t.Errorf("MaxPayloadSize = %d, want %d", p.maxPayloadSize, tt.wantMaxPayload)
			}

			if p.payloadType != RTPPayloadTypeJPEG {
				t.Errorf("PayloadType = %d, want %d", p.payloadType, RTPPayloadTypeJPEG)
			}

			if p.clockRate != RTPClockRate {
				t.Errorf("ClockRate = %d, want %d", p.clockRate, RTPClockRate)
			}
		})
	}
}

// TestPacketizeJPEG tests JPEG packetization
func TestPacketizeJPEG(t *testing.T) {
	// Create a minimal valid JPEG
	jpeg := createTestJPEG(t, 100)

	p := NewRTPPacketizer(0x12345678, 1400)

	packets, err := p.PacketizeJPEG(jpeg, 640, 480, 1000)
	if err != nil {
		t.Fatalf("PacketizeJPEG failed: %v", err)
	}

	if len(packets) == 0 {
		t.Fatal("No packets generated")
	}

	// Verify first packet
	pkt := packets[0]

	// Check RTP header
	if len(pkt) < RTPHeaderSize+JPEGHeaderSize {
		t.Fatalf("Packet too short: %d bytes", len(pkt))
	}

	// Check RTP version
	version := (pkt[0] >> 6) & 0x03
	if version != RTPVersion {
		t.Errorf("RTP version = %d, want %d", version, RTPVersion)
	}

	// Check payload type
	pt := pkt[1] & 0x7F
	if pt != RTPPayloadTypeJPEG {
		t.Errorf("Payload type = %d, want %d", pt, RTPPayloadTypeJPEG)
	}

	// Check marker bit on last packet
	lastPkt := packets[len(packets)-1]
	marker := (lastPkt[1] & 0x80) != 0
	if !marker {
		t.Error("Marker bit not set on last packet")
	}

	// Verify all packets except last don't have marker
	for i := 0; i < len(packets)-1; i++ {
		marker := (packets[i][1] & 0x80) != 0
		if marker {
			t.Errorf("Marker bit set on packet %d (not last)", i)
		}
	}

	// Check sequence numbers are sequential
	for i := 1; i < len(packets); i++ {
		seq1 := binary.BigEndian.Uint16(packets[i-1][2:4])
		seq2 := binary.BigEndian.Uint16(packets[i][2:4])

		expectedSeq := (seq1 + 1) & 0xFFFF
		if seq2 != expectedSeq {
			t.Errorf("Sequence number gap: packet %d has seq %d, expected %d", i, seq2, expectedSeq)
		}
	}

	// Verify timestamps are consistent
	ts := binary.BigEndian.Uint32(packets[0][4:8])
	for i := 1; i < len(packets); i++ {
		pktTs := binary.BigEndian.Uint32(packets[i][4:8])
		if pktTs != ts {
			t.Errorf("Timestamp mismatch in packet %d: got %d, want %d", i, pktTs, ts)
		}
	}

	// Verify SSRC
	ssrc := binary.BigEndian.Uint32(packets[0][8:12])
	if ssrc != 0x12345678 {
		t.Errorf("SSRC = %x, want %x", ssrc, 0x12345678)
	}
}

// TestPacketizeJPEGFragmentation tests large JPEG fragmentation
func TestPacketizeJPEGFragmentation(t *testing.T) {
	// Create large JPEG that requires multiple packets
	largeJPEG := createTestJPEG(t, 10000)

	p := NewRTPPacketizer(0xABCDEF00, 1400)

	packets, err := p.PacketizeJPEG(largeJPEG, 1920, 1080, 5000)
	if err != nil {
		t.Fatalf("PacketizeJPEG failed: %v", err)
	}

	// Should create multiple packets
	if len(packets) <= 1 {
		t.Errorf("Expected multiple packets for large JPEG, got %d", len(packets))
	}

	// Verify fragment offsets are increasing
	var lastOffset uint32
	for i, pkt := range packets {
		// Extract fragment offset from JPEG header (bytes 13-15)
		offset := uint32(pkt[13])<<16 | uint32(pkt[14])<<8 | uint32(pkt[15])

		if i > 0 && offset <= lastOffset {
			t.Errorf("Fragment offset not increasing: packet %d offset=%d, previous=%d", i, offset, lastOffset)
		}
		lastOffset = offset
	}

	// Verify dimensions in JPEG header
	width := int(packets[0][18]) * 8
	height := int(packets[0][19]) * 8

	if width != 1920 {
		t.Errorf("Width = %d, want 1920", width)
	}
	if height != 1080 {
		t.Errorf("Height = %d, want 1080", height)
	}
}

// TestPacketizeJPEGEmpty tests empty JPEG handling
func TestPacketizeJPEGEmpty(t *testing.T) {
	p := NewRTPPacketizer(0x12345678, 1400)

	_, err := p.PacketizeJPEG([]byte{}, 640, 480, 1000)
	if err == nil {
		t.Error("Expected error for empty JPEG, got nil")
	}
}

// TestPacketizeJPEGInvalid tests invalid JPEG handling
func TestPacketizeJPEGInvalid(t *testing.T) {
	p := NewRTPPacketizer(0x12345678, 1400)

	// Invalid JPEG (no SOI marker)
	invalidJPEG := []byte{0x00, 0x00, 0x01, 0x02, 0x03}

	_, err := p.PacketizeJPEG(invalidJPEG, 640, 480, 1000)
	if err == nil {
		t.Error("Expected error for invalid JPEG, got nil")
	}
}

// TestCalculateTimestamp tests timestamp generation
func TestCalculateTimestamp(t *testing.T) {
	p := NewRTPPacketizer(0x12345678, 1400)

	// Test 30 FPS
	ts1 := p.CalculateTimestamp(30)
	ts2 := p.CalculateTimestamp(30)

	expectedIncrement := uint32(RTPClockRate / 30)
	actualIncrement := ts2 - ts1

	if actualIncrement != expectedIncrement {
		t.Errorf("Timestamp increment = %d, want %d", actualIncrement, expectedIncrement)
	}

	// Test 15 FPS
	p.Reset()
	ts3 := p.CalculateTimestamp(15)
	ts4 := p.CalculateTimestamp(15)

	expectedIncrement = uint32(RTPClockRate / 15)
	actualIncrement = ts4 - ts3

	if actualIncrement != expectedIncrement {
		t.Errorf("Timestamp increment (15fps) = %d, want %d", actualIncrement, expectedIncrement)
	}
}

// TestSequenceNumberRollover tests sequence number wrapping
func TestSequenceNumberRollover(t *testing.T) {
	p := NewRTPPacketizer(0x12345678, 1400)

	// Set sequence number near rollover
	p.sequenceNumber = 0xFFFE

	jpeg := createTestJPEG(t, 100)

	packets, err := p.PacketizeJPEG(jpeg, 640, 480, 1000)
	if err != nil {
		t.Fatalf("PacketizeJPEG failed: %v", err)
	}

	// Get final sequence number
	lastSeq := binary.BigEndian.Uint16(packets[len(packets)-1][2:4])

	// Should have wrapped around
	if lastSeq <= 0xFFFE && len(packets) > 2 {
		t.Errorf("Sequence number didn't wrap: last seq = %d", lastSeq)
	}
}

// TestGetStats tests statistics tracking
func TestGetStats(t *testing.T) {
	p := NewRTPPacketizer(0x12345678, 1400)

	// Initial stats should be zero
	stats := p.GetStats()
	if stats.PacketsSent != 0 || stats.BytesSent != 0 || stats.FramesSent != 0 {
		t.Errorf("Initial stats not zero: %+v", stats)
	}

	jpeg := createTestJPEG(t, 500)

	// Send a few frames
	for i := 0; i < 5; i++ {
		_, err := p.PacketizeJPEG(jpeg, 640, 480, uint32(i*3000))
		if err != nil {
			t.Fatalf("PacketizeJPEG failed: %v", err)
		}
	}

	stats = p.GetStats()

	if stats.FramesSent != 5 {
		t.Errorf("FramesSent = %d, want 5", stats.FramesSent)
	}

	if stats.PacketsSent == 0 {
		t.Error("PacketsSent is zero")
	}

	if stats.BytesSent == 0 {
		t.Error("BytesSent is zero")
	}

	if stats.BytesSent != uint64(len(jpeg)*5) {
		t.Errorf("BytesSent = %d, want %d", stats.BytesSent, len(jpeg)*5)
	}
}

// TestReset tests packetizer reset
func TestReset(t *testing.T) {
	p := NewRTPPacketizer(0x12345678, 1400)

	jpeg := createTestJPEG(t, 200)

	// Send some packets
	p.PacketizeJPEG(jpeg, 640, 480, 1000)

	stats := p.GetStats()
	if stats.FramesSent == 0 {
		t.Fatal("No frames sent before reset")
	}

	// Reset
	p.Reset()

	stats = p.GetStats()
	if stats.PacketsSent != 0 || stats.BytesSent != 0 || stats.FramesSent != 0 {
		t.Errorf("Stats not reset: %+v", stats)
	}

	if p.GetSequenceNumber() != 0 {
		t.Errorf("Sequence number not reset: %d", p.GetSequenceNumber())
	}

	if p.GetNextTimestamp() != 0 {
		t.Errorf("Timestamp not reset: %d", p.GetNextTimestamp())
	}
}

// TestTimestampGenerator tests timestamp generation
func TestTimestampGenerator(t *testing.T) {
	tg := NewTimestampGenerator(30)

	// Test frame-based generation
	ts1 := tg.NextFrameBased(0)
	ts2 := tg.NextFrameBased(1)
	ts3 := tg.NextFrameBased(2)

	expectedIncrement := uint32(RTPClockRate / 30)

	if ts2-ts1 != expectedIncrement {
		t.Errorf("Frame 1 increment = %d, want %d", ts2-ts1, expectedIncrement)
	}

	if ts3-ts2 != expectedIncrement {
		t.Errorf("Frame 2 increment = %d, want %d", ts3-ts2, expectedIncrement)
	}

	// Test different FPS
	tg15 := NewTimestampGenerator(15)
	ts15_1 := tg15.NextFrameBased(0)
	ts15_2 := tg15.NextFrameBased(1)

	expectedIncrement15 := uint32(RTPClockRate / 15)
	if ts15_2-ts15_1 != expectedIncrement15 {
		t.Errorf("15fps increment = %d, want %d", ts15_2-ts15_1, expectedIncrement15)
	}
}

// TestConcurrentPacketization tests thread safety
func TestConcurrentPacketization(t *testing.T) {
	p := NewRTPPacketizer(0x12345678, 1400)
	jpeg := createTestJPEG(t, 300)

	const numGoroutines = 10
	const packetsPerGoroutine = 20

	done := make(chan bool, numGoroutines)

	for i := 0; i < numGoroutines; i++ {
		go func(id int) {
			for j := 0; j < packetsPerGoroutine; j++ {
				_, err := p.PacketizeJPEG(jpeg, 640, 480, uint32(id*1000+j*100))
				if err != nil {
					t.Errorf("Goroutine %d: PacketizeJPEG failed: %v", id, err)
				}
			}
			done <- true
		}(i)
	}

	// Wait for all goroutines
	for i := 0; i < numGoroutines; i++ {
		<-done
	}

	stats := p.GetStats()
	expectedFrames := uint64(numGoroutines * packetsPerGoroutine)

	if stats.FramesSent != expectedFrames {
		t.Errorf("FramesSent = %d, want %d", stats.FramesSent, expectedFrames)
	}
}

// TestJPEGHeaderValues tests JPEG header construction
func TestJPEGHeaderValues(t *testing.T) {
	p := NewRTPPacketizer(0x12345678, 1400)
	jpeg := createTestJPEG(t, 100)

	packets, err := p.PacketizeJPEG(jpeg, 1920, 1080, 5000)
	if err != nil {
		t.Fatalf("PacketizeJPEG failed: %v", err)
	}

	pkt := packets[0]

	// Type Specific (byte 12)
	typeSpecific := pkt[12]
	if typeSpecific != 0 {
		t.Errorf("TypeSpecific = %d, want 0", typeSpecific)
	}

	// Fragment Offset (bytes 13-15) - should be 0 for first packet
	fragmentOffset := uint32(pkt[13])<<16 | uint32(pkt[14])<<8 | uint32(pkt[15])
	if fragmentOffset != 0 {
		t.Errorf("FragmentOffset = %d, want 0 for first packet", fragmentOffset)
	}

	// Type (byte 16)
	jpegType := pkt[16]
	if jpegType != 0 {
		t.Errorf("JPEG Type = %d, want 0 (baseline)", jpegType)
	}

	// Q (byte 17)
	q := pkt[17]
	if q != 128 {
		t.Errorf("Q = %d, want 128 (dynamic)", q)
	}

	// Width (byte 18)
	width := int(pkt[18]) * 8
	if width != 1920 {
		t.Errorf("Width = %d, want 1920", width)
	}

	// Height (byte 19)
	height := int(pkt[19]) * 8
	if height != 1080 {
		t.Errorf("Height = %d, want 1080", height)
	}
}

// Helper function to create a minimal valid JPEG for testing
func createTestJPEG(t *testing.T, payloadSize int) []byte {
	t.Helper()

	var buf bytes.Buffer

	// SOI marker
	buf.Write([]byte{0xFF, 0xD8})

	// Add some payload data
	for i := 0; i < payloadSize; i++ {
		buf.WriteByte(byte(i % 256))
	}

	// EOI marker
	buf.Write([]byte{0xFF, 0xD9})

	return buf.Bytes()
}

// Benchmark tests
func BenchmarkPacketizeJPEG(b *testing.B) {
	p := NewRTPPacketizer(0x12345678, 1400)
	jpeg := createBenchmarkJPEG(5000) // 5KB typical JPEG

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, _ = p.PacketizeJPEG(jpeg, 640, 480, uint32(i*3000))
	}
}

func BenchmarkPacketizeLargeJPEG(b *testing.B) {
	p := NewRTPPacketizer(0x12345678, 1400)
	jpeg := createBenchmarkJPEG(50000) // 50KB large JPEG

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, _ = p.PacketizeJPEG(jpeg, 1920, 1080, uint32(i*3000))
	}
}

func createBenchmarkJPEG(size int) []byte {
	buf := make([]byte, size+4)
	buf[0] = 0xFF
	buf[1] = 0xD8
	buf[size+2] = 0xFF
	buf[size+3] = 0xD9
	return buf
}

File: rust-mjpeg-rtp/tests/rtp_packetizer_test.rs
//! Comprehensive tests for RTP packetizer (equivalent to Go tests)

use rust_mjpeg_rtp::rtp::{
    RtpPacketizer, TimestampGenerator, JPEG_HEADER_SIZE, RTP_CLOCK_RATE, RTP_HEADER_SIZE,
    RTP_PAYLOAD_TYPE_JPEG, RTP_VERSION,
};

/// Helper to create test JPEG with SOI/EOI markers
fn create_test_jpeg(payload_size: usize) -> Vec<u8> {
    let mut jpeg = vec![0xFF, 0xD8]; // SOI marker
    jpeg.extend((0..payload_size).map(|i| (i % 256) as u8));
    jpeg.extend(&[0xFF, 0xD9]); // EOI marker
    jpeg
}

#[test]
fn test_new_rtp_packetizer() {
    let p = RtpPacketizer::new(0x12345678, 1400);
    let stats = p.get_stats();

    assert_eq!(stats.packets_sent, 0);
    assert_eq!(stats.bytes_sent, 0);
    assert_eq!(stats.frames_sent, 0);
}

#[test]
fn test_new_packetizer_default_mtu() {
    let p = RtpPacketizer::new(0x12345678, 0);
    // Should use default MTU
    assert_eq!(p.get_stats().packets_sent, 0);
}

#[test]
fn test_packetize_jpeg_single_packet() {
    let jpeg = create_test_jpeg(100);
    let p = RtpPacketizer::new(0x12345678, 1400);

    let packets = p.packetize_jpeg(&jpeg, 640, 480, 1000).unwrap();

    assert!(!packets.is_empty());

    // Verify first packet structure
    let pkt = &packets[0];
    assert!(pkt.len() >= RTP_HEADER_SIZE + JPEG_HEADER_SIZE);

    // Check RTP version
    let version = (pkt[0] >> 6) & 0x03;
    assert_eq!(version, RTP_VERSION, "RTP version mismatch");

    // Check payload type
    let pt = pkt[1] & 0x7F;
    assert_eq!(pt, RTP_PAYLOAD_TYPE_JPEG, "Payload type mismatch");

    // Check timestamp
    let ts = u32::from_be_bytes([pkt[4], pkt[5], pkt[6], pkt[7]]);
    assert_eq!(ts, 1000, "Timestamp mismatch");

    // Check SSRC
    let ssrc = u32::from_be_bytes([pkt[8], pkt[9], pkt[10], pkt[11]]);
    assert_eq!(ssrc, 0x12345678, "SSRC mismatch");
}

#[test]
fn test_packetize_jpeg_marker_bit() {
    let jpeg = create_test_jpeg(100);
    let p = RtpPacketizer::new(0x12345678, 1400);

    let packets = p.packetize_jpeg(&jpeg, 640, 480, 1000).unwrap();

    // Last packet should have marker bit set
    let last = &packets[packets.len() - 1];
    let marker = (last[1] & 0x80) != 0;
    assert!(marker, "Marker bit not set on last packet");

    // All other packets should NOT have marker bit
    if packets.len() > 1 {
        for (i, pkt) in packets[..packets.len() - 1].iter().enumerate() {
            let marker = (pkt[1] & 0x80) != 0;
            assert!(!marker, "Marker bit set on packet {} (not last)", i);
        }
    }
}

#[test]
fn test_packetize_jpeg_fragmentation() {
    // Create large JPEG that requires multiple packets
    let large_jpeg = create_test_jpeg(10000);
    let p = RtpPacketizer::new(0xABCDEF00, 1400);

    let packets = p.packetize_jpeg(&large_jpeg, 1920, 1080, 5000).unwrap();

    // Should create multiple packets
    assert!(
        packets.len() > 1,
        "Expected multiple packets for large JPEG"
    );

    // Verify fragment offsets are increasing
    let mut last_offset = 0u32;
    for (i, pkt) in packets.iter().enumerate() {
        // Extract fragment offset from JPEG header (bytes 13-15)
        let offset = ((pkt[13] as u32) << 16) | ((pkt[14] as u32) << 8) | (pkt[15] as u32);

        if i > 0 {
            assert!(
                offset > last_offset,
                "Fragment offset not increasing at packet {}",
                i
            );
        }
        last_offset = offset;
    }
}

#[test]
fn test_packetize_jpeg_sequence_numbers() {
    let jpeg = create_test_jpeg(5000);
    let p = RtpPacketizer::new(0x12345678, 1400);

    let packets = p.packetize_jpeg(&jpeg, 640, 480, 1000).unwrap();

    if packets.len() > 1 {
        // Verify sequence numbers are sequential
        for i in 1..packets.len() {
            let seq1 = u16::from_be_bytes([packets[i - 1][2], packets[i - 1][3]]);
            let seq2 = u16::from_be_bytes([packets[i][2], packets[i][3]]);

            let expected = seq1.wrapping_add(1);
            assert_eq!(
                seq2, expected,
                "Sequence number gap at packet {}: got {}, expected {}",
                i, seq2, expected
            );
        }
    }
}

#[test]
fn test_packetize_jpeg_timestamps_consistent() {
    let jpeg = create_test_jpeg(5000);
    let p = RtpPacketizer::new(0x12345678, 1400);

    let packets = p.packetize_jpeg(&jpeg, 640, 480, 2000).unwrap();

    // All packets from same frame should have same timestamp
    let ts = u32::from_be_bytes([packets[0][4], packets[0][5], packets[0][6], packets[0][7]]);

    for (i, pkt) in packets.iter().enumerate() {
        let pkt_ts = u32::from_be_bytes([pkt[4], pkt[5], pkt[6], pkt[7]]);
        assert_eq!(pkt_ts, ts, "Timestamp mismatch at packet {}", i);
    }
}

#[test]
fn test_jpeg_header_dimensions() {
    let jpeg = create_test_jpeg(100);
    let p = RtpPacketizer::new(0x12345678, 1400);

    let packets = p.packetize_jpeg(&jpeg, 1920, 1080, 1000).unwrap();
    let pkt = &packets[0];

    // Width in 8-pixel blocks (byte 18)
    let width = pkt[18] as u32 * 8;
    assert_eq!(width, 1920, "Width mismatch");

    // Height in 8-pixel blocks (byte 19)
    let height = pkt[19] as u32 * 8;
    assert_eq!(height, 1080, "Height mismatch");
}

#[test]
fn test_jpeg_header_type_and_q() {
    let jpeg = create_test_jpeg(100);
    let p = RtpPacketizer::new(0x12345678, 1400);

    let packets = p.packetize_jpeg(&jpeg, 640, 480, 1000).unwrap();
    let pkt = &packets[0];

    // Type-specific (byte 12)
    assert_eq!(pkt[12], 0, "Type-specific should be 0");

    // Type (byte 16) - baseline JPEG
    assert_eq!(pkt[16], 0, "JPEG type should be 0 (baseline)");

    // Q (byte 17) - dynamic quantization
    assert_eq!(pkt[17], 128, "Q should be 128 (dynamic)");
}

#[test]
fn test_jpeg_header_fragment_offset_first_packet() {
    let jpeg = create_test_jpeg(100);
    let p = RtpPacketizer::new(0x12345678, 1400);

    let packets = p.packetize_jpeg(&jpeg, 640, 480, 1000).unwrap();
    let pkt = &packets[0];

    // Fragment offset (bytes 13-15) should be 0 for first packet
    let offset = ((pkt[13] as u32) << 16) | ((pkt[14] as u32) << 8) | (pkt[15] as u32);
    assert_eq!(offset, 0, "Fragment offset should be 0 for first packet");
}

#[test]
fn test_packetize_empty_jpeg() {
    let p = RtpPacketizer::new(0x12345678, 1400);
    let result = p.packetize_jpeg(&[], 640, 480, 1000);
    assert!(result.is_err(), "Should fail on empty JPEG");
}

#[test]
fn test_packetize_invalid_jpeg_no_soi() {
    let p = RtpPacketizer::new(0x12345678, 1400);
    let invalid = vec![0x00, 0x00, 0x01, 0x02, 0xFF, 0xD9];
    let result = p.packetize_jpeg(&invalid, 640, 480, 1000);
    assert!(result.is_err(), "Should fail on invalid JPEG (no SOI)");
}

#[test]
fn test_packetize_invalid_jpeg_no_eoi() {
    let p = RtpPacketizer::new(0x12345678, 1400);
    let invalid = vec![0xFF, 0xD8, 0x01, 0x02, 0x03, 0x04];
    let result = p.packetize_jpeg(&invalid, 640, 480, 1000);
    assert!(result.is_err(), "Should fail on invalid JPEG (no EOI)");
}

#[test]
fn test_sequence_number_rollover() {
    let jpeg = create_test_jpeg(100);
    let p = RtpPacketizer::new(0x12345678, 1400);

    // Manually set sequence number near rollover
    // Note: We'll send multiple frames to trigger rollover
    for i in 0..0xFFFF + 5 {
        let _ = p.packetize_jpeg(&jpeg, 640, 480, i * 3000);
    }

    let stats = p.get_stats();
    // Sequence number should have wrapped
    assert!(
        stats.current_seq < 10,
        "Sequence number should have wrapped"
    );
}

#[test]
fn test_calculate_timestamp() {
    let p = RtpPacketizer::new(0x12345678, 1400);

    // Test 30 FPS
    let ts1 = p.calculate_timestamp(30);
    let ts2 = p.calculate_timestamp(30);

    let expected_increment = RTP_CLOCK_RATE / 30;
    let actual_increment = ts2.wrapping_sub(ts1);

    assert_eq!(
        actual_increment, expected_increment,
        "Timestamp increment mismatch for 30 FPS"
    );
}

#[test]
fn test_get_stats() {
    let jpeg = create_test_jpeg(500);
    let p = RtpPacketizer::new(0x12345678, 1400);

    // Initial stats
    let stats = p.get_stats();
    assert_eq!(stats.frames_sent, 0);
    assert_eq!(stats.packets_sent, 0);
    assert_eq!(stats.bytes_sent, 0);

    // Send some frames
    for i in 0..5 {
        p.packetize_jpeg(&jpeg, 640, 480, i * 3000).unwrap();
    }

    let stats = p.get_stats();
    assert_eq!(stats.frames_sent, 5);
    assert!(stats.packets_sent > 0);
    assert_eq!(stats.bytes_sent, (jpeg.len() * 5) as u64);
}

#[test]
fn test_reset() {
    let jpeg = create_test_jpeg(200);
    let p = RtpPacketizer::new(0x12345678, 1400);

    // Send some packets
    p.packetize_jpeg(&jpeg, 640, 480, 1000).unwrap();

    let stats = p.get_stats();
    assert!(stats.frames_sent > 0);

    // Reset
    p.reset();

    let stats = p.get_stats();
    assert_eq!(stats.frames_sent, 0);
    assert_eq!(stats.packets_sent, 0);
    assert_eq!(stats.bytes_sent, 0);
    assert_eq!(stats.current_seq, 0);
    assert_eq!(stats.current_ts, 0);
}

#[test]
fn test_timestamp_generator() {
    let tg = TimestampGenerator::new(30);

    // Test frame-based generation
    let ts1 = tg.next_frame_based(0);
    let ts2 = tg.next_frame_based(1);
    let ts3 = tg.next_frame_based(2);

    let expected_increment = RTP_CLOCK_RATE / 30;

    assert_eq!(ts2 - ts1, expected_increment);
    assert_eq!(ts3 - ts2, expected_increment);
}

#[test]
fn test_timestamp_generator_different_fps() {
    let tg15 = TimestampGenerator::new(15);
    let ts1 = tg15.next_frame_based(0);
    let ts2 = tg15.next_frame_based(1);

    let expected_increment = RTP_CLOCK_RATE / 15;
    assert_eq!(ts2 - ts1, expected_increment);
}

#[test]
fn test_concurrent_packetization() {
    use std::sync::Arc;
    use std::thread;

    let p = Arc::new(RtpPacketizer::new(0x12345678, 1400));
    let jpeg = create_test_jpeg(300);

    const NUM_THREADS: usize = 10;
    const PACKETS_PER_THREAD: usize = 20;

    let mut handles = vec![];

    for i in 0..NUM_THREADS {
        let p_clone = Arc::clone(&p);
        let jpeg_clone = jpeg.clone();

        let handle = thread::spawn(move || {
            for j in 0..PACKETS_PER_THREAD {
                let ts = (i * 1000 + j * 100) as u32;
                p_clone.packetize_jpeg(&jpeg_clone, 640, 480, ts).unwrap();
            }
        });

        handles.push(handle);
    }

    for handle in handles {
        handle.join().unwrap();
    }

    let stats = p.get_stats();
    let expected_frames = (NUM_THREADS * PACKETS_PER_THREAD) as u64;

    assert_eq!(stats.frames_sent, expected_frames);
    assert_eq!(
        stats.bytes_sent,
        (jpeg.len() * expected_frames as usize) as u64
    );
}

#[test]
fn test_large_jpeg() {
    // Test with realistically large JPEG
    let large_jpeg = create_test_jpeg(100_000); // ~100KB
    let p = RtpPacketizer::new(0x12345678, 1400);

    let packets = p.packetize_jpeg(&large_jpeg, 1920, 1080, 5000).unwrap();

    // Should create many packets
    assert!(packets.len() > 50, "Expected many packets for 100KB JPEG");

    // Verify total payload size
    let total_payload: usize = packets
        .iter()
        .map(|p| p.len() - RTP_HEADER_SIZE - JPEG_HEADER_SIZE)
        .sum();

    assert_eq!(total_payload, large_jpeg.len());
}
